tcms/testruns/static/testruns/js/get.js

```diff
@@ -85,9 +86,48 @@ function renderProgressBars(positiveCount, negativeCount, allCount) {
 
 function renderCountPerStatusList(statusCount) {
     for (var status in statusCount) {
-        const status_id = statusCount[status].id;
+        const statusId = statusCount[status].id;
 
-        $(`#count-for-status-${status_id}`).attr('href', `?status_id=${status_id}`);
-        $(`#count-for-status-${status_id}`).text(statusCount[status].count);
+        $(`#count-for-status-${statusId}`).attr('href', `?status_id=${statusId}`).text(statusCount[status].count);
     }
 }
+
+function renderTestExecutions(testRunId) {
+    const container = $('#test-executions-container')
+    const testExecutionRowTemplate = $('#test-execution-row')[0].content
+
+    jsonRPC('TestExecution.filter', { 'run_id': testRunId }, testExecutions => {
+        testExecutions.forEach(testExecution => {
+            jsonRPC('TestCase.filter', { 'id': testExecution.case_id }, testCases => {
+                jsonRPC('TestExecutionStatus.filter', { 'id': testExecution.status_id }, statuses => {
+                    jsonRPC('Bug.filter', {'execution_id': testExecution.pk}, bugs => {
+                        const template = $(testExecutionRowTemplate.cloneNode(true))
+                        container.append(renderTestExecutionRow(template, testExecution, testCases[0], statuses[0], bugs))
+
+                        treeViewBind();
```

Comment:
`treeViewBind` should not be executed for every item rendered, no ? Between 110-111 looks good to me.

Reply:
the way it is now is called only once - after the `testExecutions.forEach` loop has finished. putting it outside of the `jsonRPC` callback would mean that it will be invoked before all the executions have been rendered, and the tree view will not be binded at all.

Reply:
yes, that was wrong. now it should be ok.

Reply:
actually, the reason I put it there in the first place is because we need to be sure all executions are rendered before calling `treeViewBind`, therefore - 840a494

Reply:
now both firefox and chrome complain:
```
[Deprecation] Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience. For more help, check https://xhr.spec.whatwg.org/.
```
and that is one more reason to refactor `jsonRPC` to return promises or something else that handles situations like these properly

Reply:
Refactoring the code to promises instead of callbacks is a completely different story. Doesn't have anything to do here.

> actually, the reason I put it there in the first place is because we need to be sure al

I agree. Upon a second look the placement of `treeViewBind()` is exactly where it is supposed to be. 

However the fact that you have nested RPC calls (which by default are async) is the reason there may be race condition between rendering each row of TE and calling treeViewBind(). You try to overcome this by making the nested API call syncronous, which browsers don't like b/c the main thread will block the rest of the page loading process.

The underlying reason for all of this is that you need TestCase category & priority for rendering. Bugs count, also for rendering. 

One possible solution is to skip rendering Category, Priority & Bug count initially; render the tree view; then make another pass to render the needed values. You can make use of the fact that you have all case IDs already so you can pass them like `pk__in: [1, 2, 3, 4]` instead of fetching the data row by row.

Reply:
> Refactoring the code to promises instead of callbacks is a completely different story. Doesn't have anything to do here.

I know this is out of scope and we are not going to fix it now, just proving a point.

I like your proposed solution, because now we are using async functions the way they are meant to be used. However, that applies only to testcases, because there the query is `execution_id__in`. 

However, with bugs we have a different story, because there the relation is many to many, so we should query them execution by execution. If I am missing something, or you have a better proposal, I am all ears.

Reply:
For Bug.filter you can do `executions__in=[list of ids]` and you will get all bugs for all TEs at once. However then you will have to group them by TE then count how many there are. Let's leave it like that for now. Supposedly most of the time there won't be that many bugs in the TR.